### PR TITLE
Allow for test program compilation to override CFLAGS

### PR DESCRIPTION
**Pull request by danielgustafsson**
In case a test program requires a specialized set of CFLAGS, allow it to be overriden with `LOCAL_CFLAGS` and `LOCAL_CFLAGS_FILTER`. Also include an override from the tcop tests where -O3 is causing problems with post-ereport assertions (reported to -dev in https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/sFMpzprBWG4). By going with -O0 we can instruct the compiler to not remove unreachable code which allows the test to perform assertions after the ereport.

### CODE REVIEW

**Comment by danielgustafsson** [comment]
So Concourse claims that the test is still failing with this, but some insights from the more Concourse savvy devs would be nice. Looking at the output in the failing build I don't see the unittests being compiled, they appear to run without compilation? Is Concourse hiding the STDOUT output for the unittest compilation or is the tree not cleaned before building so that the binaries remain?  If the latter is true then this needs to be fixed with a `make distclean` (or other means of having a fresh tree) IMHO, else we won't be able to test things like Makefile patches.

File: src/backend/tcop/test/Makefile [hunk 1]
```diff
@@ -2,6 +2,12 @@ subdir=src/backend/tcop
 top_builddir=../../../..
 include $(top_builddir)/src/Makefile.global
 
+# Override CFLAGS optimization level with -O0 since a more aggressive level
+# will optimize on the unreachable code information which in turns breaks
+# the assertions performed after the ereport({FATAL|ERROR} ...); call.
+LOCAL_CFLAGS_FILTER := -O3
+LOCAL_CFLAGS := -O0
```
Comment (asimrp) [inline src/backend/tcop/test/Makefile]: Wow, optimization turned out to be the final culprit.  Thank you for the analysis and the apt comment.

**Comment by asimrp** [comment]
I tried this PR on my OSX.  The tcop test continues to fail.  Looking at the compile command, I still see the O3 flag there.
```
apraveen ~/gpdb/src/backend/tcop/test]$ git show -s 
commit 07c8dbd7bf1487a74c0bb25853487acf38753018
Author: Daniel Gustafsson <dgustafsson@pivotal.io>
Date:   Thu Sep 15 12:28:04 2016 +0200

    Allow for test program compilation to override CFLAGS
    
    In case a test program requires a specialized set of CFLAGS it can
    be overriden with LOCAL_CFLAGS and LOCAL_CFLAGS_FILTER.
    
    Also include an override from the tcop tests where -O3 is causing
    problems with post-ereport assertions.
[apraveen ~/gpdb/src/backend/tcop/test]$
[apraveen ~/gpdb/src/backend/tcop/test]$ make
gcc -no-cpp-precomp -O3 -std=gnu99  -Wall -Wmissing-prototypes -Wpointer-arith -Winline -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -Wno-address -g -ggdb  -w -pthread  -D_REENTRANT -D_THREAD_SAFE -D_POSIX_PTHREAD_SEMANTICS -I../../../../src/include -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/usr/include/libxml2  -I../../../../src/test/unit/cmockery -I../../../../src/backend/libpq -I../../../../src/backend/gp_libpq_fe -I../../../../src/backend/postmaster -I. -I../../../../src/port -DDLSUFFIX=.so -I../../../../src/backend/utils/stat  -c -o postgres_test.o postgres_test.c
gcc -no-cpp-precomp -O3 -std=gnu99  -Wall -Wmissing-prototypes -Wpointer-arith -Winline -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -Wno-address -g -ggdb  -w -pthread  -D_REENTRANT -D_THREAD_SAFE -D_POSIX_PTHREAD_SEMANTICS -I../../../../src/include -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/usr/include/libxml2  -I../../../../src/test/unit/cmockery -I../../../../src/backend/libpq -I../../../../src/backend/gp_libpq_fe -I../../../../src/backend/postmaster -I. -I../../../../src/port -DDLSUFFIX=.so -I../../../../src/backend/utils/stat  -c -o ../../../../src/test/unit/mock/backend/utils/error/elog_mock.o ../../../../src/test/unit/mock/backend/utils/error/elog_mock.c

   ...

[apraveen ~/gpdb/src/backend/tcop/test]$ ./postgres.t 
[ RUN         ] test__IsTransactionExitStmtList__MultipleElementList
[          OK ] test__IsTransactionExitStmtList__MultipleElementList
[ RUN         ] test__IsTransactionExitStmt__IsTransactionStmt
[          OK ] test__IsTransactionExitStmt__IsTransactionStmt
[ RUN         ] test__IsTransactionExitStmt__IsQuery
[          OK ] test__IsTransactionExitStmt__IsQuery
[ RUN         ] test__ProcessInterrupts__ClientConnectionLost
Abort trap: 6
[apraveen ~/gpdb/src/backend/tcop/test]$
```

**Comment by d** [comment]
Hey Daniel,
Thanks for the good work for keeping master green(er). We are in the process of reproducing / verifying the fix. But before that, what we can tell you is: Concourse does not have incremental / dirty builds. Every build is a complete clean build. Hope this helps clarify things.

Thanks,
Bhuvnesh and Jesse

**Comment by d** [comment]
Hey Daniel,
We ran `make unittest-check -C src/backend` and see the same unit tests failing. A snippet of failure message is enclosed below:

> [ RUN         ] test__IsTransactionExitStmtList__MultipleElementList
> [          OK ] test__IsTransactionExitStmtList__MultipleElementList
> [ RUN         ] test__IsTransactionExitStmt__IsTransactionStmt
> [          OK ] test__IsTransactionExitStmt__IsTransactionStmt
> [ RUN         ] test__IsTransactionExitStmt__IsQuery
> [          OK ] test__IsTransactionExitStmt__IsQuery
> [ RUN         ] test__ProcessInterrupts__ClientConnectionLost
>                 No entries for symbol errmsg.
>                 ERROR: ../../../../src/test/unit/mock/backend/utils/error/elog_mock.c:338 - Could not get value to check parameter fmt of function errmsg
>                 Previously declared parameter value was declared at postgres_test.c:67
> [      FAILED ] test__ProcessInterrupts__ClientConnectionLost
> [ RUN         ] test__ProcessInterrupts__DoingCommandRead
>                 No entries for symbol errmsg.
>                 ERROR: ../../../../src/test/unit/mock/backend/utils/error/elog_mock.c:338 - Could not get value to check parameter fmt of function errmsg
>                 Previously declared parameter value was declared at postgres_test.c:143
> [      FAILED ] test__ProcessInterrupts__DoingCommandRead
> [=============] 5 tests ran
> [ PASSED      ] 3 tests
> [ FAILED      ] 2 tests, listed below
> [ FAILED      ] test__ProcessInterrupts__ClientConnectionLost
> [ FAILED      ] test__ProcessInterrupts__DoingCommandRead

We noticed that your change only changes the optimization flag (substituting `-O0` for `-O3`) for `postgres.t` but the mock and test object files are still compiled with `-O3`, is that intentional?

For a little context, in our repro, the flags used for `postgres_test.o` was:
```
-O3 -std=gnu99  -Wall -Wmissing-prototypes -Wpointer-arith -Winline -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -Wno-unused-but-set-variable -Wno-address  -w -pthread  -D_REENTRANT -D_THREAD_SAFE -D_POSIX_PTHREAD_SEMANTICS
```

whereas the flags for `postgres.t` was
```
-std=gnu99 -Wall -Wmissing-prototypes -Wpointer-arith -Winline -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -Wno-unused-but-set-variable -Wno-address -w -pthread -D_REENTRANT -D_THREAD_SAFE -D_POSIX_PTHREAD_SEMANTICS -O0
```

We don't have enough context on this, but we'd like to help you move along and get our builds back green. Hope this helps.

Thanks,
Jesse and Bhuvnesh

**Comment by asimrp** [comment]
On Thu, Sep 15, 2016 at 12:22 PM, Jesse Zhang <notifications@github.com>
wrote:

>
> For a little context, in our repro, the flags used for postgres_test.o
> was:
>
> -O3 -std=gnu99  -Wall -Wmissing-prototypes -Wpointer-arith -Winline -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -Wno-unused-but-set-variable -Wno-address  -w -pthread  -D_REENTRANT -D_THREAD_SAFE -D_POSIX_PTHREAD_SEMANTICS
>
>
>
And I bet the flags used for elog_mock.o includes `-O3` as well.  It
happened in my tests too.  In fact, my `configure` arguments include
`CFLAGS=-O0 CPPFLAGS=-O0`.  I didn't dig through, however, to see where
`-O3` is coming from.

**Comment by danielgustafsson** [comment]
So it turns out I was tricked by a non-clean tree, I clearly had too little ☕ during hacking and missed that. Given the structure of our Makefiles it turned messy to extend this to actually handle optimization override where we need it so I abandoned the attempt and instead borrowed a page from the playbook of another unittest for error handling. By stealing errfinish and doing a siglongjmp we can regain control. This passes for me on a completely fresh tree so fingers crossed Concourse is happy with me now. Sorry for the noise.

**Comment by danielgustafsson** [comment]
@d: I suspected Concourse did fresh builds (anything else is dangerous IMHO) but I'm still puzzled as to why the compile output from the unittest is hidden? At least I can't see where the .t and the _mock.o files are compiled in the concourse output, am I looking in the wrong place?

File: src/backend/tcop/test/postgres_test.c [hunk 2]
```diff
@@ -135,18 +171,21 @@ test__ProcessInterrupts__DoingCommandRead(void **state)
...
-	expect_any(errstart, lineno);
-	expect_any(errstart, funcname);
-	expect_any(errstart, domain);
-	will_return(errstart, false);
-
 	will_be_called(DisableNotifyInterrupt);
 	will_be_called(DisableCatchupInterrupt);
 
-	ProcessInterrupts();
+	EXPECT_EREPORT(ERROR);
```
Comment (asimrp) [inline src/backend/tcop/test/postgres_test.c]: Nitpick - for log level of `ERROR`, we don't need to mock `errfinish()`, the real `errfinish()` will also call `siglongjmp()` the same way.  I'm throwing this comment to simply verify my understanding, it doesn't imply any code change.

**Comment by asimrp** [comment]
+1 :+1:

**Comment by xinzweb** [comment]
@danielgustafsson Just expand `build_with_orca`, and you should see the unittests ran.

**Comment by danielgustafsson** [comment]
@xinzweb I see the the tests running, just not compiling which I find odd since it's a normal compilation step echoing to STDOUT just like the rest of GPDB compilation. If I need to hijack something it indicates that it's actively hidden, I'm simply trying to understand why.

**Comment by d** [comment]
Daniel,
The standard behavior of `make` is that for each target, the command outputs are printed. It's implementation-defined behavior whether the command being run _per se_ is printed out.

In the case of your first patch set of this PR, yes it would have been nice to see the commands. Although we could also have reasoned it through your 10-line code change, right?

**Comment by danielgustafsson** [comment]
Rebased and pushed to master, thanks for review!

**Pull request closed without merging**